### PR TITLE
Redo the majority of Certbot's pinning system

### PR DESCRIPTION
**Pull request by bmw**
Fixes https://github.com/certbot/certbot/issues/8705, https://github.com/certbot/certbot/issues/8548, and https://github.com/certbot/certbot/issues/8290.

This implements the majority of what's described under the "Core Need and Idea" section of https://docs.google.com/document/d/12AP7JfY_uq0QAt-dwqdUX4NYUtjQ-MmDTS8ldoLI38I/edit?usp=sharing. I think the only thing that's missing from what I described there is I'd like to write more documentation. I kept that out for now though because:

1. If we implement something similar for the oldest tests as described in that doc, things like the scripts we use could change significantly and I'd rather not waste my time.
2. I think/hope the comments at the top of `requirements.txt` and `pin.sh` is enough for now.

In general, I personally think that what I have here is pretty robust, however, there are many, many things we could/should do to improve things. I wanted to get a simple version of this up for review though and I think we can continue to iterate on it as desired.

You can see all tests passing on this PR at https://dev.azure.com/certbot/certbot/_build/results?buildId=3667&view=results.

### CODE REVIEW

**Comment by bmw** [comment]
I'll take a look at the `test_sdists` failure which I thought I fixed tomorrow. I think we're close here though!

**Review by adferrand — CHANGES_REQUESTED** [review]
Except a version range for `poetry` to be consistent, this LGTM. I think that you can also rerun `pin.sh` to keep the most up-to-date dependencies that are not included yet in the PR.

File: tools/pinning/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+version = "0.1.0"
+description = "A simple project for pinning Certbot's dependencies using Poetry."
+authors = ["Certbot Project"]
+license = "Apache License 2.0"
+
+[tool.poetry.dependencies]
+python = "^3.6"
+
+# Local dependencies
+# Any local packages that have dependencies on other local packages must be
```
[the remark this answers is not in the document]
Comment (bmw) [reply tools/pinning/pyproject.toml:12]: Yeah I honestly don't entirely understand this, but it seems to work and I also tested it when our local packages rely on unreleased versions and everything still seems to work fine.

If you think this is something I should open a poetry issue about, let me know.

**Comment by bmw** [comment]
I reran `pin.sh` and it's almost certainly overkill but to be safe and make sure everything still works with the new versions I kicked off a run of the full test suite at https://dev.azure.com/certbot/certbot/_build/results?buildId=3672&view=results.

**Review by adferrand — APPROVED** [review]
LGTM!

**Comment by alexzorin** [comment]
Did cryptography get stealth-bumped to `3.4.7` in this PR? 

Just curious if anything else needed to be done for that OpenSSL CVE yesterday, on platforms that bundled it, like Windows.

**Comment by adferrand** [comment]
For Windows there is a pre-compile wheel with `abi3` compatibility (https://files.pythonhosted.org/packages/2f/be/91f1495687c2dab970abd95868d03aa65fd766d4dd4f2204df7e8af7f1ba/cryptography-3.4.7-cp36-abi3-win32.whl), so it does not need to be compiled and so no `rust` compiler is needed.

There were only an issue to compile for armhf/arm64 architectures, but it has been fixed with https://github.com/certbot/certbot/pull/8730.

**Comment by bmw** [comment]
> Did cryptography get stealth-bumped to 3.4.7 in this PR?

I didn't mean to bump anything stealthily, but yes, the pinning did get updated to `cryptography==3.4.7` in this PR. I'm going to write up documentation on all this soon, but in general, our pinning file is now simply `tools/requirements.txt`.

> Just curious if anything else needed to be done for that OpenSSL CVE yesterday, on platforms that bundled it, like Windows.

For the record, the OpenSSL CVEs are at https://www.openssl.org/news/vulnerabilities.html#y2021.

Bumping `cryptography` resolves any potential problems using precompiled `cryptography` wheels as long as this PR is included, but we have to worry about any versions of OpenSSL we bundle outside of that. I checked and our snap and Docker sources have updated so we will automatically get the latest version there when the images are rebuilt.

For Windows though, Python has not put out a new build using OpenSSL 1.1.1k. See https://bugs.python.org/issue43631. Inspecting the `libssl-1_1.dll` included in the version of Python we use on Windows, it looks like it's using OpenSSL 1.1.1i which is affected.

For now I'm planning to keep an eye on that issue and delaying our release/bumping up our release/investigating the CVEs more based on their response. If someone thinks we should do something else, please speak up!

**Comment by bmw** [comment]
If you followed the thread on the Python bug tracker, it unfortunately doesn't look like Python is going to update their vendored version of OpenSSL anytime soon 🙁 

Looking at the CVEs though and pulling heavily from https://mail.python.org/archives/list/python-dev@python.org/thread/2GULUR43MNEW3IJM44LS5ZY2TOUANPNT/, I don't think the CVEs affect modern versions of Certbot. I think it could affect our acme library's support for tls-alpn-01, but this code isn't exposed to Certbot users.

Because of all this, I'm not planning to do anything special so we'd release new snaps and Docker images using OpenSSL 1.1.1k next week and the Windows version of Certbot will continue to use an older version of OpenSSL for now.

**Pull request merged**